### DISCUSSION TITLE
edgedb error: ConstraintViolationError: may not make non-empty object type 'default::Person' abstract

### DISCUSSION THREAD

**Discussion by passion8** [geldata/gel#3472] [Q&A] [answered]
I am reading a book, I am on chapter 2  https://www.edgedb.com/easy-edgedb/chapter2

when I read `scalar type Transport extending enum<Feet, Train, HorseDrawnCarriage>;` , I made appropriate changes into the schema (below is the schema)

     module default {
        
        abstract type Person {
          required property name -> str;
          multi link places_visited -> City;
        }
        
        scalar type Transport extending enum<Feet, Train, HorseDrawnCarriage>;
      
        type PC extending Person {
          required property transport -> Transport;
        }
      
        type NPC extending Person {
        }
      
        type City {
          required property name -> str;
          property modern_name -> str;
          property important_places -> array<str>;
        }
      
      }

after then I ran `edgedb migration create` and ` edgedb migrate` it is throwing error

    edgedb error:  ConstraintViolationError: may not make non-empty object type 'default::Person' abstract

What does it mean? How to solve it?

I am attaching migrations, which probably might help you to debug. 

[migrations.zip](https://github.com/edgedb/edgedb/files/8055076/migrations.zip)

**Comment by elprans** [comment, marked as answer]
Turning a non-abstract `Person` type into an `abstract` one while it contains data is not allowed, because abstract types are not supposed to contain data.  

Two ways to fix this:

1. Migrate to an empty schema first.
2. Remove all data from `Person` before migrating: `delete Person` (though this may fail if there are links to person objects).

@vpetrovykh, can we do something here to remove this snag?

P/S @passion8, also please direct future book bugs to the edgedb/easy-edgedb repo.  Thanks!

**Comment by dotlouis** [comment]
Just in case someone stumbles upon this while migrating, here's a relevant discord thread to help: https://discord.com/channels/841451783728529451/1009476546085523496/1009499950054396026
